main.go

```diff
@@ -202,3 +215,37 @@ func wrapAuthHandler(handler http.HandlerFunc, htpasswdFile string, host string)
 	authenticator := auth.NewBasicAuthenticator(host, secrets)
 	return auth.JustCheck(authenticator, handler)
 }
+
+func writePidFile(c *config.Config, accessLogger cache.Logger) error {
+	// create a "pid" directory under cache directory and clean up inactive pid file
+	pidPath := filepath.Join(c.Dir, "pid")
+	err := os.MkdirAll(pidPath, os.FileMode(0744))
+	if err != nil {
+		log.Fatal(err)
+		return err
+	}
+	err = filepath.Walk(pidPath, func(name string, fileInfo os.FileInfo, err error) error {
+		if !fileInfo.IsDir() {
+			pid, _ := strconv.Atoi(fileInfo.Name())
+			bazelRemoteProcess, err := os.FindProcess(pid)
+			if err != nil {
+				accessLogger.Printf("Error to find pid: %d", bazelRemoteProcess)
+			} else {
+				err = bazelRemoteProcess.Signal(syscall.Signal(0))
+				if err != nil && strings.Contains(err.Error(), "process already finished") {
+					accessLogger.Printf("Removing pid file: %s", fileInfo.Name())
+					err = os.Remove(filepath.Join(c.Dir, "pid", fileInfo.Name()))
+				} else if c.KillOldPid {
+					accessLogger.Printf("Killing existing bazel remote process: %d", pid)
+					err = bazelRemoteProcess.Signal(syscall.Signal(syscall.SIGKILL))
```

Comment:
`SIGKILL` does not terminate the server gracefully. See [this article](https://medium.com/@kpbird/golang-gracefully-stop-application-23c2390bb212) for ways to do that.

Reply:
That's a good point. I'll see how we can better deal with it.